grpc-client-utils/src/main/java/org/hypertrace/core/grpcutils/client/GrpcChannelRegistry.java

```diff
@@ -50,7 +51,37 @@ private String getChannelId(String host, int port, boolean isPlaintext) {
   }
 
   public void shutdown() {
-    channelMap.values().forEach(ManagedChannel::shutdown);
+    channelMap.forEach(this::fullyShutdownChannel);
```

Comment:
If there were N channels, it would take worst case 2*N minutes (1 min to wait for graceful shutdown + 1 min wait for forceful shutdown) for all the channels to shutdown

Should we use parallelism to shutdown channels, instead of sequential shutting down each of them, since `awaitTermination` is a thread blocking call?

Reply:
I thought about that, but decided it felt like overkill. I'd argue this is a case where we have to look at the expected runtime vs the theoretical worst to save on complexity.

When does a shutdown hang? When there's a pending call, but we're implicitly given a longer deadline for later channels. That is, each successive channel gets the time all previous took plus another minute to complete. So the graceful part of the worst case scenario would require - N channels all with pending calls with (effectively) infinite deadlines. That's not really a scenario we have or ever should have - we always use deadlines. Now the forceful part, this should be independent of deadline since it cancels pending calls, too. My understanding is the only reason force shutdown is not instantaneous is threading - that is, the shutdown command is passed to the executor, which on complete will mark termination. So this should be quick in any scenario.

Now what happens if the worst comes to pass - every channel is in use with an infinite deadline? It should take the container N minutes to shut down, because they'd wait for 1 minute each, then forcefully shut down. So the container takes a long time to shutdown - in a k8s world, no big deal - a new replica should have been brought up and traffic diverted with this one shutting down ( we should go clean up health checks as a separate effort, btw). In a docker/compose world, it would mean it's taking a long time to shutdown and perhaps we get a little forceful - but 


Anyway, now that I've argued against it, I'll propose the following changes because there's no real harm in doing so. I still think it's overkill to bring in parallelism here, but let's call this proposal... batching :). Do you feel like the following changes would address your concerns:
- Commence all shutdowns before waiting for any
- Have a single wait period that checks all channels (we either write the wait part ourselves or share a common deadline)
- Come back through with a forceful shutdown on any unterminated channels
- Again, single (quicker - say 5s) wait for all channels.

Reply:
> That's not really a scenario we have or ever should have - we always use deadlines

We actually don't always use deadlines (_though I know we should_)

Reply:
> a new replica should have been brought up and traffic diverted with this one shutting down

I am guessing you mean, with this one terminating down successfully

Reply:
I think the proposal makes sense, if it's easier to implement. Mostly, the single wait period across all channels

Reply:
I combined the last two since it's a quick wait and force shutdown should be close to instant, but otherwise updated.

Reply:
> We actually don't always use deadlines (_though I know we should_)

please raise issues where we don't

> I am guessing you mean, with this one terminating down successfully

Not suer I phrased it correctly but no not successfully. In that scenario, a new instance would be spun up. Once it is healthy and traffic is cut over to the new instance, the old would commence shutdown. Bad hygiene, but we don't _really_ care how long it takes to do this unless as avinash pointed out, we're using stateful sets or the like.